### PR TITLE
JAMES-2708 Linshare client

### PR DESCRIPTION
**Pull request by trantienduchn**

### CODE REVIEW

File: third-party/linshare/pom.xml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        </dependency>
+        <dependency>
+            <groupId>io.github.openfeign</groupId>
+            <artifactId>feign-core</artifactId>
+            <version>${feign.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>io.github.openfeign.form</groupId>
+            <artifactId>feign-form</artifactId>
+            <version>3.8.0</version>
```
Comment (chibenwa) [inline third-party/linshare/pom.xml]: Define it in root pom please

File: third-party/linshare/pom.xml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+    <dependencies>
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.github.openfeign</groupId>
+            <artifactId>feign-core</artifactId>
+            <version>${feign.version}</version>
```
Comment (chibenwa) [inline third-party/linshare/pom.xml]: version not needed

File: third-party/linshare/pom.xml [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        </dependency>
+        <dependency>
+            <groupId>io.github.openfeign.form</groupId>
+            <artifactId>feign-form</artifactId>
+            <version>3.8.0</version>
+        </dependency>
+        <dependency>
+            <groupId>io.github.openfeign</groupId>
+            <artifactId>feign-jackson</artifactId>
+            <version>${feign.version}</version>
```
Comment (chibenwa) [inline third-party/linshare/pom.xml]: version not needed

File: third-party/linshare/pom.xml [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        </dependency>
+        <dependency>
+            <groupId>io.github.openfeign</groupId>
+            <artifactId>feign-slf4j</artifactId>
+            <version>${feign.version}</version>
```
Comment (chibenwa) [inline third-party/linshare/pom.xml]: version not needed

File: third-party/linshare/pom.xml [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.github.openfeign</groupId>
+            <artifactId>feign-core</artifactId>
+            <version>${feign.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>io.github.openfeign.form</groupId>
```
Comment (chibenwa) [inline third-party/linshare/pom.xml]: GroupId io.github.openfeign.form goes after all groupIds io.github.openfeign

File: third-party/linshare/pom.xml [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.github.steveash.guavate</groupId>
+            <artifactId>guavate</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.james</groupId>
```
Comment (chibenwa) [inline third-party/linshare/pom.xml]: use the constant + goes first.

File: third-party/linshare/src/main/java/org/apache/james/linshare/LinshareConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        @FunctionalInterface
+        public interface RequireUrl {
+            RequireUsername url(URL url);
+
+            default RequireUsername urlAsString(String url) throws MalformedURLException {
+                return url(new URL(url));
+            }
+        }
+
+        public interface RequireUsername {
```
Comment (chibenwa) [inline third-party/linshare/src/main/java/org/apache/james/linshare/LinshareConfiguration.java]: We should require a long lived token.

https://ci.linagora.com/linagora/lgs/linshare/products/linshare-github/blob/master/documentation/EN/configuration/how-to-use-jwt.md

File: third-party/linshare/src/main/java/org/apache/james/linshare/client/LinshareAPI.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    static LinshareAPI from(LinshareConfiguration configuration) {
+        LinshareCredential credential = configuration.getCredential();
+        return Feign.builder()
+            .requestInterceptor(new BasicAuthRequestInterceptor(credential.getUsername(), new String(credential.getPassword())))
+            .logger(new Slf4jLogger(LinshareAPI.class))
+            .logLevel(Logger.Level.FULL)
+            .encoder(new JacksonEncoder())
+            .encoder(new FormEncoder(new JacksonEncoder()))
+            .decoder(new JacksonDecoder())
+            .retryer(new Retryer.Default())
```
Comment (chibenwa) [inline third-party/linshare/src/main/java/org/apache/james/linshare/client/LinshareAPI.java]: Is that `retryer` really needed ? - rabbitMQ retry API is a bit specific and fails...

File: third-party/linshare/src/main/java/org/apache/james/linshare/client/LinshareAPI.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+public interface LinshareAPI {
+
+    static LinshareAPI from(LinshareConfiguration configuration) {
+        LinshareCredential credential = configuration.getCredential();
+        return Feign.builder()
+            .requestInterceptor(new BasicAuthRequestInterceptor(credential.getUsername(), new String(credential.getPassword())))
+            .logger(new Slf4jLogger(LinshareAPI.class))
+            .logLevel(Logger.Level.FULL)
+            .encoder(new JacksonEncoder())
```
Comment (trantienduchn) [inline third-party/linshare/src/main/java/org/apache/james/linshare/client/LinshareAPI.java]: remove this line

File: third-party/linshare/src/main/java/org/apache/james/linshare/client/ShareRequest.java [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+            return new ShareRequest(recipientsBuilder.build(), documentsBuilder.build());
+        }
+    }
+
+    static Builder builder() {
+        return new Builder();
+    }
+
+    private final List<Recipient> recipients;
+    private final List<DocumentId> documents;
```
Comment (trantienduchn) [inline third-party/linshare/src/main/java/org/apache/james/linshare/client/ShareRequest.java]: s/documents/documentIds

File: third-party/linshare/src/main/java/org/apache/james/linshare/LinshareConfiguration.java [hunk 11]
```diff
@@ -32,19 +32,15 @@
...
             }
         }
 
-        public interface RequireUsername {
-            RequirePassword username(String username);
-        }
-
-        public interface RequirePassword {
-            ReadyToBuild password(char[] password);
+        public interface RequireAuthorizationToken {
```
Comment (trantienduchn) [inline third-party/linshare/src/main/java/org/apache/james/linshare/LinshareConfiguration.java]: s/RequireAuthorizationToken/LinshareCredential

File: third-party/linshare/src/test/java/org/apache/james/linshare/client/LinshareAPITest.java [hunk 12]
```diff
@@ -50,48 +54,74 @@
...
+    interface AuthorizationAPI {
+
+        static AuthorizationAPI from(String linshareUrl, UsernamePassword usernamePassword) throws Exception {
+            String base64Token = Base64.getEncoder()
+                .encodeToString((usernamePassword.getUsername() + ":" + usernamePassword.getPassword())
+                .getBytes(StandardCharsets.UTF_8));
+
+            LinshareConfiguration configuration = LinshareConfiguration.builder()
+                .urlAsString(linshareUrl)
+                .credential(LinshareCredential.basic(new LinshareCredential.Token(base64Token)))
```
Comment (trantienduchn) [inline third-party/linshare/src/test/java/org/apache/james/linshare/client/LinshareAPITest.java]: in order to retrieve jwt token, I still need to use basic authentication. Then, LinshareCredential should be generic enough to deal with Basic & JWT usages

File: third-party/linshare/src/test/java/org/apache/james/linshare/client/LinshareAPITest.java [hunk 13]
```diff
@@ -50,48 +54,74 @@
...
     private LinshareAPI user2LinshareAPI;
-    private LinshareConfiguration configuration;
 
     @BeforeEach
     void setup(DockerComposeContainer<?> linshare) throws Exception {
-        configuration = LinshareConfiguration.builder()
-            .urlAsString("http://127.0.0.1:" + linshare.getServicePort(LinshareExtension.LINSHARE_BACKEND_SERVICE, LinshareExtension.LINSHARE_BACKEND_PORT))
-            .username(CREDENTIAL_USER_1.getUsername())
-            .password(CREDENTIAL_USER_1.getPassword())
+        System.out.println();
```
Comment (trantienduchn) [inline third-party/linshare/src/test/java/org/apache/james/linshare/client/LinshareAPITest.java]: remove this line

File: third-party/linshare/src/test/resources/ssl/id_rsa [hunk 14]
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Comment (trantienduchn) [inline third-party/linshare/src/test/resources/ssl/id_rsa]: I put these files to make linshare be able to generate jwt

File: third-party/linshare/src/test/resources/docker-compose.yml [hunk 15]
```diff
@@ -47,6 +47,7 @@ services:
         command: mongod --smallfiles
 
     smtp:
+        container_name: linshare_smtp
```
Comment (trantienduchn) [inline third-party/linshare/src/test/resources/docker-compose.yml]: I have to keep it, cause It makes linshare be able to reach to smtp server by the host name. otherwise, I'm fucked up

Comment (chibenwa) [inline third-party/linshare/src/test/resources/docker-compose.yml]: This name may be also given through a link defined in linshare-backend IMO 

(allowing to run this dockerfile two time at the same time and avoids naming conflicts)

**Comment by chibenwa** [comment]
```
Progress [ERROR] COMPILATION ERROR : 
[6573b5a1be9777e4bdf4451d3306f9ea5e324a3d] ([INFO] -------------------------------------------------------------
[6573b5a1be9777e4bdf4451d3306f9ea5e324a3d] 1): 5.7/9.3 MB
[ERROR] /james-project/backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/RabbitMQManagementAPI.java:[159,19] no suitable constructor found for RetryableException(java.lang.String,feign.Request.HttpMethod,java.util.Date)
[6573b5a1be9777e4bdf4451d3306f9ea5e324a3d]     constructor feign.RetryableException.RetryableException(int,java.lang.String,feign.Request.HttpMethod,java.lang.Throwable,java.util.Date) is not applicable
[6573b5a1be9777e4bdf4451d3306f9ea5e324a3d]       (actual and formal argument lists differ in length)
[6573b5a1be9777e4bdf4451d3306f9ea5e324a3d]     constructor feign.RetryableException.RetryableException(int,java.lang.String,feign.Request.HttpMethod,java.util.Date) is not applicable
[6573b5a1be9777e4bdf4451d3306f9ea5e324a3d]       (actual and formal argument lists differ in length)
[6573b5a1be9777e4bdf4451d3306f9ea5e324a3d] P[INFO] 1 error
```

File: third-party/linshare/src/main/java/org/apache/james/linshare/client/LinshareAPI.java [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+    @RequestLine("GET /linshare/webservice/rest/user/v2/documents")
+    @Headers("Accept: application/json")
+    List<Document> listAllDocuments();
+
+    @RequestLine("POST /linshare/webservice/rest/user/v2/documents")
+    @Headers({
+        "Accept: application/json",
+        "Content-Type: multipart/form-data"
+    })
+    Document uploadDocument(@Param("file") File file, @Param("filesize") long fileSize);
```
Comment (chibenwa) [inline third-party/linshare/src/main/java/org/apache/james/linshare/client/LinshareAPI.java]: What we truely want is an `InputStream`, and not a file.

Is there some ways to make this work with an input stream?

File: third-party/linshare/src/main/java/org/apache/james/linshare/client/LinshareAPI.java [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+public interface LinshareAPI {
+
+    @RequestLine("GET /linshare/webservice/rest/user/v2/documents")
+    @Headers("Accept: application/json")
+    List<Document> listAllDocuments();
+
+    @RequestLine("POST /linshare/webservice/rest/user/v2/documents")
+    @Headers({
+        "Accept: application/json",
+        "Content-Type: multipart/form-data"
```
Comment (chibenwa) [inline third-party/linshare/src/main/java/org/apache/james/linshare/client/LinshareAPI.java]: These headers are duplicated every where and should at the very least be in some constants...

File: third-party/linshare/src/main/java/org/apache/james/linshare/client/LinshareAPIGenerator.java [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+import feign.Feign;
+import feign.Logger;
+import feign.RequestInterceptor;
+import feign.RequestTemplate;
+import feign.form.FormEncoder;
+import feign.jackson.JacksonDecoder;
+import feign.jackson.JacksonEncoder;
+import feign.slf4j.Slf4jLogger;
+
+class LinshareAPIGenerator {
```
Comment (chibenwa) [inline third-party/linshare/src/main/java/org/apache/james/linshare/client/LinshareAPIGenerator.java]: LinshareApiFactory?

File: third-party/linshare/src/test/java/org/apache/james/linshare/client/LinshareAPITest.java [hunk 19]
```diff
@@ -0,0 +1,275 @@
...
+    @VisibleForTesting
+    interface LinshareUserAPI {
+
+        @RequestLine("GET /linshare/webservice/rest/user/v2/users")
+        @Headers("Accept: application/json")
+        List<User> allUsers();
+    }
+
+    @VisibleForTesting
+    interface AuthorizationAPI {
```
Comment (chibenwa) [inline third-party/linshare/src/test/java/org/apache/james/linshare/client/LinshareAPITest.java]: Please move this in a standalone class. I bet it is going to be useful...

File: third-party/linshare/src/test/java/org/apache/james/linshare/client/LinshareAPITest.java [hunk 20]
```diff
@@ -0,0 +1,275 @@
...
+
+        static AuthorizationAPI from(String linshareUrl, UsernamePassword usernamePassword) throws Exception {
```
Comment (chibenwa) [inline third-party/linshare/src/test/java/org/apache/james/linshare/client/LinshareAPITest.java]: String linshareUrl -> URL linshareUrl?

File: third-party/linshare/src/main/java/org/apache/james/linshare/client/LinshareAPIGenerator.java [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+        }
+
+        @Override
+        public void apply(RequestTemplate template) {
+            template.header("Authorization", authorizationToken);
+        }
+    }
+
+    @VisibleForTesting
+    static <T> T from(LinshareConfiguration configuration, Class<T> apiClass) {
```
Comment (chibenwa) [inline third-party/linshare/src/main/java/org/apache/james/linshare/client/LinshareAPIGenerator.java]: Isn't the class always LinshareAPI ?

Comment (trantienduchn) [reply third-party/linshare/src/main/java/org/apache/james/linshare/client/LinshareAPIGenerator.java]: it could be AuthorizationAPI or UserAPI

File: third-party/linshare/src/main/java/org/apache/james/linshare/client/ReceivedShare.java [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+        return downloaded;
+    }
+
+    @Override
+    public final boolean equals(Object o) {
+        if (o instanceof ReceivedShare) {
+            ReceivedShare that = (ReceivedShare) o;
+
+            return Objects.equals(this.sender, that.sender)
+                && Objects.equals(this.document, that.document);
```
Comment (trantienduchn) [inline third-party/linshare/src/main/java/org/apache/james/linshare/client/ReceivedShare.java]: missed downloaded property

File: third-party/linshare/src/test/java/org/apache/james/linshare/client/LinshareAPITest.java [hunk 23]
```diff
@@ -0,0 +1,205 @@
...
+    void listAllShouldReturnAllUploadedDocuments() throws Exception {
+        Document firstDocument = user1LinshareAPI.uploadDocument(templateFile());
+        Document secondDocument = user1LinshareAPI.uploadDocument(templateFile());
+
+        assertThat(user1LinshareAPI.listAllDocuments())
+            .containsExactly(firstDocument, secondDocument);
+    }
+
+    @Test
+    void listAllShouldNotReturnAllDocumentsOfOtherUsers() throws Exception {
```
Comment (trantienduchn) [inline third-party/linshare/src/test/java/org/apache/james/linshare/client/LinshareAPITest.java]: s/NotReturnAllDocuments/NotReturnDocuments

File: third-party/linshare/src/main/java/org/apache/james/linshare/client/LinshareAPI.java [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+    @Headers(ACCEPT_APPLICATION_JSON)
+    List<Document> listAllDocuments();
+
+    @RequestLine("POST /linshare/webservice/rest/user/v2/documents")
+    @Headers({ ACCEPT_APPLICATION_JSON, CONTENT_TYPE_MULTIPART })
+    Document uploadDocument(@Param("file") File file, @Param("filesize") long fileSize);
+
+    default Document uploadDocument(File file) {
+        Preconditions.checkNotNull(file);
+        Preconditions.checkArgument(file.exists(), "upload file is not existed");
```
Comment (chibenwa) [inline third-party/linshare/src/main/java/org/apache/james/linshare/client/LinshareAPI.java]: `"File to upload does not exist: " + fileURL`

File: third-party/linshare/src/test/java/org/apache/james/linshare/LinshareExtension.java [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+import feign.form.FormEncoder;
+import feign.jackson.JacksonDecoder;
+import feign.jackson.JacksonEncoder;
+import feign.slf4j.Slf4jLogger;
+
+public class LinshareExtension implements BeforeAllCallback, BeforeEachCallback, AfterAllCallback {
+
+    private interface LinshareAPIForTesting {
+
+        class JWTToken {
```
Comment (chibenwa) [inline third-party/linshare/src/test/java/org/apache/james/linshare/LinshareExtension.java]: This IMO should be `AuthorizationToken`...

File: third-party/linshare/src/test/java/org/apache/james/linshare/LinshareExtension.java [hunk 26]
```diff
@@ -0,0 +1,133 @@
...
+    private Linshare linshare;
+
+    @Override
+    public void beforeAll(ExtensionContext context) throws Exception {
+        linshare = Linshare.create();
+        linshare.start();
+    }
+
+    @Override
+    public void beforeEach(ExtensionContext context) throws Exception {
```
Comment (chibenwa) [inline third-party/linshare/src/test/java/org/apache/james/linshare/LinshareExtension.java]: This should be done in afterEach as this is a cleanup step...

File: third-party/linshare/src/test/java/org/apache/james/linshare/LinshareFixture.java [hunk 27]
```diff
@@ -0,0 +1,83 @@
...
+
+import org.testcontainers.shaded.com.google.common.collect.ImmutableList;
+
+import com.github.steveash.guavate.Guavate;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+
+public interface LinshareFixture {
+
+    class UsernamePassword {
```
Comment (chibenwa) [inline third-party/linshare/src/test/java/org/apache/james/linshare/LinshareFixture.java]: This can be named `Credentials` IMO

File: third-party/linshare/src/main/java/org/apache/james/linshare/AuthorizationToken.java [hunk 28]
```diff
@@ -0,0 +1,57 @@
...
+
+    private final String value;
+
+    @VisibleForTesting
+    public AuthorizationToken(String value) {
+        Preconditions.checkArgument(StringUtils.isNotBlank(value), "token value cannot be null or blank");
+        this.value = value;
+    }
+
+    public String asString() {
```
Comment (chibenwa) [inline third-party/linshare/src/main/java/org/apache/james/linshare/AuthorizationToken.java]:
```
public String asBearerHeader() {
   return "Bearer: " + value;
}
```

File: backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/RabbitMQManagementAPI.java [hunk 29]
```diff
@@ -156,7 +156,7 @@ static RabbitMQManagementAPI from(RabbitMQConfiguration configuration) {
 
     ErrorDecoder RETRY_500 = (methodKey, response) -> {
         if (response.status() == 500) {
-            throw new RetryableException("Error encountered, scheduling retry", response.request().httpMethod(), new Date());
+            throw new RetryableException(response.status(), "Error encountered, scheduling retry", response.request().httpMethod(), new Date());
```
Comment (rouazana) [inline backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/RabbitMQManagementAPI.java]: unrelated change?

**Comment by Arsnael** [comment]
Read it !

File: third-party/linshare/src/test/resources/docker-compose.yml [hunk 30]
```diff
@@ -16,9 +16,11 @@ services:
         links:
             - database:linshare_database
             - mongodb:linshare_mongodb
+            - smtp:linshare_smtp
```
Comment (rouazana) [inline third-party/linshare/src/test/resources/docker-compose.yml]: you cannot use james smtp?

File: third-party/linshare/src/main/java/org/apache/james/linshare/client/LinshareAPI.java [hunk 31]
```diff
@@ -0,0 +1,115 @@
...
+import feign.RequestLine;
+import feign.RequestTemplate;
+import feign.form.FormEncoder;
+import feign.jackson.JacksonDecoder;
+import feign.jackson.JacksonEncoder;
+import feign.slf4j.Slf4jLogger;
+
+public interface LinshareAPI {
+
+    interface Header {
```
Comment (chibenwa) [inline third-party/linshare/src/main/java/org/apache/james/linshare/client/LinshareAPI.java]: Headers

File: third-party/linshare/src/main/java/org/apache/james/linshare/client/LinshareAPI.java [hunk 32]
```diff
@@ -0,0 +1,115 @@
...
+        String ACCEPT_APPLICATION_JSON = "Accept: application/json";
+        String CONTENT_TYPE_MULTIPART = "Content-Type: multipart/form-data";
+        String CONTENT_TYPE_APPLICATION_JSON = "Content-Type: application/json";
+    }
+
+    class AuthorizationInterceptor implements RequestInterceptor {
+
+        private final String authorizationToken;
```
Comment (chibenwa) [inline third-party/linshare/src/main/java/org/apache/james/linshare/client/LinshareAPI.java]: Should be of type AuthorizationToken (let's keep the strong types to the edge of our code)

File: third-party/linshare/src/test/java/org/apache/james/linshare/AuthorizationTokenTest.java [hunk 33]
```diff
@@ -0,0 +1,49 @@
...
+        assertThatThrownBy(() -> new AuthorizationToken(nullToken))
+            .isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    void constructorShouldThrowWhenTokenIsBlank() {
+        String blankToken = "    ";
+        assertThatThrownBy(() -> new AuthorizationToken(blankToken))
+            .isInstanceOf(IllegalArgumentException.class);
+    }
```
Comment (chibenwa) [inline third-party/linshare/src/test/java/org/apache/james/linshare/AuthorizationTokenTest.java]: Please add a test for `asHeaderValue`

File: third-party/linshare/src/test/java/org/apache/james/linshare/LinshareConfigurationTest.java [hunk 34]
```diff
@@ -0,0 +1,52 @@
...
+    @Test
+    void constructorShouldThrowWhenPassingNullCredential() {
+        AuthorizationToken nullToken = null;
+        assertThatThrownBy(() -> new LinshareConfiguration(new URL("https://linshare.linagora.com"), nullToken))
+            .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    void constructorShouldThrowWhenPassingNullUrl() {
+        AuthorizationToken token = new AuthorizationToken("Bearer jwt-token-at-here");
```
Comment (chibenwa) [inline third-party/linshare/src/test/java/org/apache/james/linshare/LinshareConfigurationTest.java]: Remove the `Bearer` prefix here

File: backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/RabbitMQManagementAPI.java (hunk 29 above)
Comment (chibenwa) [reply backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/RabbitMQManagementAPI.java]: No I bet duc did upgrade feign version (was a compilation error he did encountered)

File: third-party/linshare/src/test/resources/docker-compose.yml (hunk 30 above)
Comment (chibenwa) [reply third-party/linshare/src/test/resources/docker-compose.yml]: This third party project do not depend on a james product hence this would technically be impossible - at least as long as we are not writing integration tests - not the case here.

Also, we will aim at checking that the side effect we hope for (sharing with linshare) is performed. We can ensure this by browsing the sharing API.

We would thus never need a real configuration for the smtp - we don't care about LinShare internal mechanism IMO. So I would argue that even on integration tests this might not be desirable. Let's not enforce too much LinShare internal mechanisms...

Also please note that having a dockerized service connecting to a host service might not be trivial, the ip address affected to the James server varying... Better keep things simple...

File: third-party/linshare/src/test/java/org/apache/james/linshare/LinshareDockerCompose.java [hunk 35]
```diff
@@ -0,0 +1,34 @@
...
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.linshare;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+
+interface LinshareDockerCompose {
```
Comment (chibenwa) [inline third-party/linshare/src/test/java/org/apache/james/linshare/LinshareDockerCompose.java]: Why is this interface needed? What does it add?

File: third-party/linshare/src/test/java/org/apache/james/linshare/SingletonLinshare.java [hunk 36]
```diff
@@ -0,0 +1,55 @@
...
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.linshare;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+public class SingletonLinshare implements LinshareDockerCompose {
```
Comment (chibenwa) [inline third-party/linshare/src/test/java/org/apache/james/linshare/SingletonLinshare.java]: LisshareSingletong & it just need to hold a static linshare, it don't need to **be** a static linShare

**Comment by chibenwa** [comment]
test this please

File: third-party/linshare/src/test/java/org/apache/james/linshare/LinshareDockerCompose.java (hunk 35 above)
Comment (trantienduchn) [reply third-party/linshare/src/test/java/org/apache/james/linshare/LinshareDockerCompose.java]: What I want is provide an Linshare Interface to the client, where it can be a normal linshare or singleton linshare. With singleton linshare, I can override start() & stop() to do nothing while with normal linshare, it acts start() & stop() normally. 

Clients can have a choice to choose which impl it interests while still be able to use a same suit of apis.

**Comment by Arsnael** [comment]
Read it

**Comment by trantienduchn** [comment]
test this please

File: third-party/linshare/src/test/java/org/apache/james/linshare/Linshare.java [hunk 37]
```diff
@@ -30,6 +30,14 @@
     private static final String LINSHARE_BACKEND_SERVICE = "backend";
     private static final int LINSHARE_BACKEND_PORT = 8080;
 
+    public static Linshare createSneakyThrow() {
```
Comment (mbaechler) [inline third-party/linshare/src/test/java/org/apache/james/linshare/Linshare.java]: it's not sneakyThrow it's just runtime wrapping

**Comment by hoangdat** [comment]
Read it

File: pom.xml [hunk 38]
```diff
@@ -525,18 +525,6 @@
...
-        <module>examples</module>
-        <module>javax-mail-extension</module>
-        <module>mailbox</module>
-        <module>mailet</module>
-        <module>mdn</module>
-        <module>metrics</module>
-        <module>mpt</module>
-        <module>protocols</module>
-        <module>server</module>
         <module>third-party</module>
```
Comment (chibenwa) [inline pom.xml]: Oooups

File: third-party/linshare/src/test/java/org/apache/james/linshare/Linshare.java [hunk 39]
```diff
@@ -0,0 +1,136 @@
...
+            .withEnv("MONGODB_PORT", "27017")
+            .withEnv("THUMBNAIL_ENABLE", "false")
+            .withExposedPorts(LINSHARE_BACKEND_PORT)
+            .waitingFor(Wait.forLogMessage(WAIT_FOR_BACKEND_INIT_LOG, 1)
+                .withStartupTimeout(Duration.ofMinutes(10)))
+            .withNetwork(network);
+    }
+
+    public String getUrl() {
+        return "http://127.0.0.1:" + getPort();
```
Comment (chibenwa) [inline third-party/linshare/src/test/java/org/apache/james/linshare/Linshare.java]: Won't work on the CI

Prefer: `"http://" + getHost() + ":" + getPort();`

Comment (chibenwa) [inline third-party/linshare/src/test/java/org/apache/james/linshare/Linshare.java]: Won't work on the CI

Prefer: `"http://" + getHost() + ":" + getPort();`

**Comment by chibenwa** [comment]
Just rebased

**Comment by chibenwa** [comment]
```
[db3b9a42bc2d55d207004999bddb8c5fc0cbdee8] docker: Error response from daemon: Cannot link to a non running container: /elasticSearch-linshare-client-23215ea1-45e6-419c-b0d2-fbb8e1f83529 AS /james-server-cassandra-rabbitmq-linshare-client-23215ea1-45e6-419c-b0d2-fbb8e1f83529/elasticsearch.
```

Compil + mvn test -> OK

An error on deployment tests

test this please
